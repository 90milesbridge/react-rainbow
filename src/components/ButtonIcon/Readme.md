ButtonIcon examples:

    <div>
        <div style={{padding: '0.5rem'}}>
            <ButtonIcon iconName="utility:activity" size="large" />
        </div>
        <div style={{padding: '0.5rem', backgroundColor: 'rgb(244, 246, 249)'}}>
            <ButtonIcon iconName="utility:adduser" variant="border-filled" size="small" />
        </div>
        <div style={{padding: '0.5rem', backgroundColor: 'rgb(22, 50, 92)'}}>
            <ButtonIcon iconName="utility:automate" variant="border-inverse" size="large" />
        </div>
        <div style={{padding: '0.5rem', backgroundColor: 'rgb(244, 246, 249)'}}>
            <ButtonIcon iconName="utility:broadcast" variant="border" size="x-small" />
        </div>
        <div style={{padding: '0.5rem'}}>
            <ButtonIcon iconName="utility:search" variant="brand" />
        </div>
        <div style={{padding: '0.5rem'}}>
            <ButtonIcon iconName="utility:search" variant="brand" disabled />
        </div>
        <div style={{padding: '0.5rem', backgroundColor: 'rgb(22, 50, 92)'}}>
            <ButtonIcon iconName="utility:dialing" variant="inverse" />
        </div>
        <div style={{padding: '0.5rem', backgroundColor: 'rgb(244, 246, 249)'}}>
            <ButtonIcon iconName="utility:chart" variant="container" size="xx-small" />
        </div>
    </div>
