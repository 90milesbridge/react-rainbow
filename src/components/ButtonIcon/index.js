import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import IconSvg from '../IconSvg';
import AssistiveText from './../AssistiveText';

/**
 * Buttons Icons provide the user with visual iconography that
 * is typically used to invoke an event or action.
 */
export default class ButtonIcon extends Component {
    constructor(props) {
        super(props);
        this.click = this.click.bind(this);
        this.focus = this.focus.bind(this);
        this.blur = this.blur.bind(this);
    }

    getIconSizeClassName() {
        const { size, variant } = this.props;
        if (variant === 'base') {
            return {
                'slds-button__icon_large': size === 'large',
                'slds-button__icon_small': size === 'small',
                'slds-button__icon_x-small': size === 'x-small' || size === 'xx-small',
            };
        }
        return null;
    }

    getIconClassNames() {
        return classnames('slds-button__icon', this.getIconSizeClassName());
    }

    getButtonSizeClassName() {
        const { size, variant } = this.props;
        if (variant === 'base') {
            return null;
        }
        return {
            'slds-button_icon-small': size === 'small',
            'slds-button_icon-x-small': size === 'x-small',
            'slds-button_icon-xx-small': size === 'xx-small',
        };
    }

    getVariantClassName() {
        const { variant } = this.props;
        if (variant === 'base') {
            return null;
        }
        return `slds-button_icon-${variant}`;
    }

    getButtonClassNames() {
        const { className } = this.props;
        return classnames('slds-button', 'slds-button_icon', this.getVariantClassName(), this.getButtonSizeClassName(), className);
    }

    /**
    * Sets focus on the element.
    *
    * @public
    */
    focus() {
        this.button.focus();
    }

    /**
    * Sets click on the element.
    *
    * @public
    */
    click() {
        this.button.click();
    }

    /**
    * Sets blur on the element.
    *
    * @public
    */
    blur() {
        this.button.blur();
    }

    render() {
        const {
            style,
            disabled,
            tabIndex,
            onFocus,
            onBlur,
            onClick,
            title,
            type,
            ariaHaspopup,
            iconName,
            assistiveText,
        } = this.props;

        return (
            <button
                data-id="button-icon-element"
                className={this.getButtonClassNames()}
                style={style}
                disabled={disabled}
                tabIndex={tabIndex}
                onFocus={onFocus}
                onBlur={onBlur}
                onClick={onClick}
                title={title}
                type={type}
                aria-haspopup={ariaHaspopup}
                ref={(ref) => { this.button = ref; }} >
                <IconSvg iconName={iconName} className={this.getIconClassNames()} />
                <AssistiveText text={assistiveText} />
            </button>
        );
    }
}

ButtonIcon.propTypes = {
    /** The class name of the root element. */
    className: PropTypes.string,
    /** It is an object with custom style applied to the root element. */
    style: PropTypes.object,
    /** Indicates that the element has a popup context menu or sub-level menu. It is
     used for srceen readers. */
    ariaHaspopup: PropTypes.bool,
    /** The variant for the button with predefined styles. */
    variant: PropTypes.oneOf([
        'base',
        'border-filled',
        'border-inverse',
        'border',
        'brand',
        'inverse',
        'container',
    ]),
    /** Disables the button if set to true. */
    disabled: PropTypes.bool,
    /** Tab index */
    tabIndex: PropTypes.number,
    /** Event fired when the button is clicked. */
    onClick: PropTypes.func,
    /** Event fired when the button is focused. */
    onFocus: PropTypes.func,
    /** Event fired when the button is blurred. */
    onBlur: PropTypes.func,
    /** This is a description that is showed when a user hover the button. */
    title: PropTypes.string,
    /** The type of the button. */
    type: PropTypes.oneOf([
        'button', 'submit', 'reset',
    ]),
    /** The name of the icon. Names are written in the
     format '\utility:down\' where 'utility' is the category, and 'down' is the
     specific icon to be displayed. */
    iconName: PropTypes.string.isRequired,
    /** Description for assistive sreen readers */
    assistiveText: PropTypes.string,
    /** The button icon size. */
    size: PropTypes.oneOf([
        'xx-small',
        'x-small',
        'small',
        'medium',
        'large',
    ]),
};

ButtonIcon.defaultProps = {
    className: undefined,
    style: {},
    ariaHaspopup: false,
    variant: 'base',
    disabled: false,
    tabIndex: undefined,
    onClick: () => {},
    onFocus: () => {},
    onBlur: () => {},
    title: undefined,
    type: 'button',
    assistiveText: undefined,
    size: 'medium',
};
