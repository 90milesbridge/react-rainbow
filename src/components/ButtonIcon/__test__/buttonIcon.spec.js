import React from 'react';
import { mount } from 'enzyme/build/index';
import renderer from 'react-test-renderer';
import ButtonIcon from './../index';

describe('<ButtonIcon/>', () => {
    it('should not render the assistive text when is not passed', () => {
        const component = mount(
            <ButtonIcon />,
        );
        expect(component.find('.slds-assistive-text').length).toBe(0);
    });
    it('should render the assistive text passed', () => {
        const component = mount(
            <ButtonIcon assistiveText="for screen readers" />,
        );
        expect(component.find('.slds-assistive-text').text()).toBe('for screen readers');
    });
    it('should focus the button when the focus method is called', () => {
        const component = mount(
            <ButtonIcon />,
        );

        component.instance().focus();
        const focusedElementDataId = document.activeElement.getAttribute('data-id');
        const buttonDataId = component.find('button').prop('data-id');
        expect(focusedElementDataId).toBe(buttonDataId);
    });
    it('should blur the button when the blur method is called', () => {
        const component = mount(
            <ButtonIcon />,
        );
        const instance = component.instance();
        const buttonDataId = component.find('button').prop('data-id');

        instance.focus();
        expect(document.activeElement.getAttribute('data-id')).toBe(buttonDataId);
        instance.blur();
        expect(document.activeElement.getAttribute('data-id')).toBeNull();
    });
    it('should set disable to true when it is passed', () => {
        const component = mount(
            <ButtonIcon disabled />,
        );

        expect(component.find('button').prop('disabled')).toBe(true);
    });
    it('should set the tabIndex passed', () => {
        const component = mount(
            <ButtonIcon tabIndex={0} />,
        );

        expect(component.find('button').prop('tabIndex')).toBe(0);
    });
    it('should set the title passed', () => {
        const component = mount(
            <ButtonIcon title="my title" />,
        );

        expect(component.find('button').prop('title')).toBe('my title');
    });
    it('should set the type passed', () => {
        const component = mount(
            <ButtonIcon type="reset" />,
        );

        expect(component.find('button').prop('type')).toBe('reset');
    });
    it('should set aria-haspopup to true when it is passed', () => {
        const component = mount(
            <ButtonIcon ariaHaspopup />,
        );

        expect(component.find('button').prop('aria-haspopup')).toBe(true);
    });
    it('should have the right class names', () => {
        const component = renderer.create(
            <ButtonIcon />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant border-filled and have a custom class', () => {
        const component = renderer.create(
            <ButtonIcon variant="border-filled" className="my-custom-class-name" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant border-inverse and size small', () => {
        const component = renderer.create(
            <ButtonIcon variant="border-inverse" size="small" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant border and size x-small', () => {
        const component = renderer.create(
            <ButtonIcon variant="border" size="x-small" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant brand and size xx-small', () => {
        const component = renderer.create(
            <ButtonIcon variant="brand" size="xx-small" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant inverse', () => {
        const component = renderer.create(
            <ButtonIcon variant="inverse" />,
        );
        expect(component).toMatchSnapshot();
    });
    it('should have the right class names when variant container', () => {
        const component = renderer.create(
            <ButtonIcon variant="container" />,
        );
        expect(component).toMatchSnapshot();
    });
});
