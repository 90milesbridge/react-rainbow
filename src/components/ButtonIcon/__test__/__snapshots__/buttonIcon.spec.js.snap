// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonIcon/> should have the right class names 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
/>
`;

exports[`<ButtonIcon/> should have the right class names when variant is base and size is large 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  <svg
    aria-hidden={true}
    className="slds-button__icon slds-button__icon_large"
    style={Object {}}
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#activity"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</button>
`;

exports[`<ButtonIcon/> should have the right class names when variant is base and size is medium 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  <svg
    aria-hidden={true}
    className="slds-button__icon"
    style={Object {}}
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#activity"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</button>
`;

exports[`<ButtonIcon/> should have the right class names when variant is base and size is small 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  <svg
    aria-hidden={true}
    className="slds-button__icon slds-button__icon_small"
    style={Object {}}
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#activity"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</button>
`;

exports[`<ButtonIcon/> should have the right class names when variant is base and size is x-small 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  <svg
    aria-hidden={true}
    className="slds-button__icon slds-button__icon_x-small"
    style={Object {}}
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#activity"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</button>
`;

exports[`<ButtonIcon/> should have the right class names when variant is base and size is xx-small 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
>
  <svg
    aria-hidden={true}
    className="slds-button__icon slds-button__icon_x-small"
    style={Object {}}
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#activity"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</button>
`;

exports[`<ButtonIcon/> should have the right class names when variant is border and size is x-small 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border slds-button_icon-x-small"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
/>
`;

exports[`<ButtonIcon/> should have the right class names when variant is border-filled and have a custom class 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border-filled my-custom-class-name"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
/>
`;

exports[`<ButtonIcon/> should have the right class names when variant is border-inverse and size is small 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-border-inverse slds-button_icon-small"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
/>
`;

exports[`<ButtonIcon/> should have the right class names when variant is brand and size is xx-small 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-brand slds-button_icon-xx-small"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
/>
`;

exports[`<ButtonIcon/> should have the right class names when variant is container 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-container"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
/>
`;

exports[`<ButtonIcon/> should have the right class names when variant is inverse and size is large 1`] = `
<button
  aria-haspopup={false}
  className="slds-align_absolute-center slds-button slds-button_icon slds-button_icon-inverse"
  data-id="button-icon-element"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={Object {}}
  type="button"
/>
`;
